File: python_solutions/1072.flip-columns-for-maximum-number-of-equal-rows.cpp
/*
 * @lc app=leetcode id=1072 lang=cpp
 *
 * [1072] Flip Columns For Maximum Number of Equal Rows
 *
 * https://leetcode.com/problems/flip-columns-for-maximum-number-of-equal-rows/description/
 *
 * algorithms
 * Medium (48.64%)
 * Total Accepted:    2.6K
 * Total Submissions: 5K
 * Testcase Example:  '[[0,1],[1,1]]'
 *
 * Given a matrix consisting of 0s and 1s, we may choose any number of columns
 * in the matrix and flip every cell in that column.  Flipping a cell changes
 * the value of that cell from 0 to 1 or from 1 to 0.
 *
 * Return the maximum number of rows that have all values equal after some
 * number of flips.
 *
 *
 *
 *
 *
 *
 *
 * Example 1:
 *
 *
 * Input: [[0,1],[1,1]]
 * Output: 1
 * Explanation: After flipping no values, 1 row has all values equal.
 *
 *
 *
 * Example 2:
 *
 *
 * Input: [[0,1],[1,0]]
 * Output: 2
 * Explanation: After flipping values in the first column, both rows have equal
 * values.
 *
 *
 *
 * Example 3:
 *
 *
 * Input: [[0,0,0],[0,0,1],[1,1,0]]
 * Output: 2
 * Explanation: After flipping values in the first two columns, the last two
 * rows have equal values.
 *
 *
 *
 *
 * Note:
 *
 *
 * 1 <= matrix.length <= 300
 * 1 <= matrix[i].length <= 300
 * All matrix[i].length's are equal
 * matrix[i][j] is 0 or 1
 *
 *
 *
 *
 */
using namespace std;
using LL = long long;
using uLL = unsigned long long;
using VI = vector<int>;
#define EACH(i, n) for (int i = 0; i <= int(n); ++i)
#define EACHV(i, n) for (int i = int(n); i >= 0; --i)
#define unfold(i, arr) for (auto &i : arr)

class Solution {
public:
  int maxEqualRowsAfterFlips(vector<vector<int>> &matrix) {
    unordered_map<uLL, int> has;
    int res = 0;
    unfold(row, matrix) res = max(res, ++has[toKey(row)]);
    return res;
  }

  uLL toKey(VI &row) {
    bool iv = row[0] == 1;
    uLL key = 0;
    unfold(n, row) {
      if (iv)
        n = !n;
      key = (key << 1) | n;
    }
    return key;
  }

  int maxEqualRowsAfterFlips_2(vector<vector<int>> &matrix) {
    int m = matrix.size(), n = matrix[0].size(), res = 0;
    EACH(i, m - 1) {
      VI flip(matrix[i]);
      int cur = 0;
      EACH(j, n - 1) flip[j] = 1 - flip[j];
      EACH(j, m - 1) if (matrix[j] == matrix[i] || flip == matrix[j]) cur += 1;
      res = max(cur, res);
    }
    return res;
  }
};

static const int _ = []() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();

File: python_solutions/1072.flip-columns-for-maximum-number-of-equal-rows.js
var bisect_left = function(nums, target) {
    var i = 0,
        j = nums.length - 1,
        mid;
    while (i < j) {
        mid = ~~((i + j) / 2);
        if (nums[mid] < target)
            i = mid + 1;
        else
            j = mid;
    }
    return i;
}

var bisect_right = function(nums, target) {
    var i = 0,
        j = nums.length - 1,
        mid;
    while (i < j) {
        mid = Math.ceil((i + j) / 2);
        if (nums[mid] > target)
            j = mid - 1;
        else
            i = mid;
    }
    return j;
}

var two_d_array = function(m, n) {
    return new Array(m).fill(0).map(() => new Array(n).fill(0));
}

var reverse = function(s) {
    return [...str].reverse().join('');
}


// simple print
var print = function(a) {
    console.log(a);
}

// For counting elements in a list 
var counter = function(array_or_string) {
    var dict = {};
    for (let i = 0; i < array_or_string.length; i++) {
        let e = array_or_string[i];
        dict[e] = dict[e] ? dict[e] += 1 : 1;
    }
    return dict;
}

var sum = function(array) {
    return array.reduce(function(a, b) {
        return a + b;
    }, 0);
}

var isodd = function(n) {
    return n % 2 == 1;
}

var iseven = function(n) {
    return n % 2 == 0;
}

var list = function(size, value) {
    return Array(size).fill(value);
}


var len = function(str_or_array) {
    return str_or_array.length;
}

var sorted = function(arr) {
    return arr.sort((a, b) => (a - b));
}

// print out a Map 
function printMap(m) {
    m.forEach(function(v, k) {
        console.log(k + " " + v);
    })
}

var gcd = function(a, b) {
    return b == 0 ? a : gcd(b, a % b);
}

var say = function(a) {
    console.log(a)
};


var two_d_array = function(m, n, v) {
    return [...Array(m)].map(x => Array(n).fill(v));
}

function floor(float_number) {
    return Math.floor(float_number)
}

function ceil(float_number) {
    return Math.ceil(float_number)
}


var max = function(arr) {
    return Math.max(...arr);
}

var min = function(arr) {
    return Math.min(...arr)
};

var pairmax = function(a, b) {
    return Math.max(a, b)
}

var pairmin = function(a, b) {
    return Math.min(a, b)
}

/*
 * @lc app=leetcode id=1072 lang=javascript
 *
 * [1072] Flip Columns For Maximum Number of Equal Rows
 *
 * https://leetcode.com/problems/flip-columns-for-maximum-number-of-equal-rows/description/
 *
 * algorithms
 * Medium (48.64%)
 * Total Accepted:    2.6K
 * Total Submissions: 5K
 * Testcase Example:  '[[0,1],[1,1]]'
 *
 * Given a matrix consisting of 0s and 1s, we may choose any number of columns
 * in the matrix and flip every cell in that column.  Flipping a cell changes
 * the value of that cell from 0 to 1 or from 1 to 0.
 * 
 * Return the maximum number of rows that have all values equal after some
 * number of flips.
 * 
 * 
 * 
 * 
 * 
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: [[0,1],[1,1]]
 * Output: 1
 * Explanation: After flipping no values, 1 row has all values equal.
 * 
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: [[0,1],[1,0]]
 * Output: 2
 * Explanation: After flipping values in the first column, both rows have equal
 * values.
 * 
 * 
 * 
 * Example 3:
 * 
 * 
 * Input: [[0,0,0],[0,0,1],[1,1,0]]
 * Output: 2
 * Explanation: After flipping values in the first two columns, the last two
 * rows have equal values.
 * 
 * 
 * 
 * 
 * Note:
 * 
 * 
 * 1 <= matrix.length <= 300
 * 1 <= matrix[i].length <= 300
 * All matrix[i].length's are equal
 * matrix[i][j] is 0 or 1
 * 
 * 
 * 
 * 
 */
/**
 * @param {number[][]} matrix
 * @return {number}
 */

var maxEqualRowsAfterFlips = function(matrix) {
    var m = new Map(),
        res = 0;
    matrix.forEach(function(row) {
        var s = "";
        for (let i = 0; i < len(row); i++)
            s += row[0] == 0 ? row[i] : (1 - row[i]);
        m[s] = m[s] ? m[s] + 1 : 1;
        res = pairmax(m[s], res);
    })
    return res;
};


var matrix = [
    [0, 0, 0],
    [0, 0, 1],
    [1, 1, 0]
];
print(maxEqualRowsAfterFlips(matrix));

var n = 8; 
var s = "";
s += n;
print(n.toString())
print(s)

